### ISSUE TITLE
"uv init --package" modifies the package name

### ISSUE THREAD

**Issue by fconil** [astral-sh/uv#12046]
### Summary

Hi,

I want to use the same name for "package name" and "import package".

Using different names confuses me and may cause security issues.

> In October 2022 [sklearn](https://pypi.org/project/sklearn/) downloads were about 1/5 of the [scikit-learn](https://pypi.org/project/scikit-learn/) downloads on PyPI so a lot of people are using it

If I try to use an underscore in the package name :

```bash
$ uv init --package a_b
Initialized project `a-b` at `/home/fconil/Progs/python/packaging/uv-0.6.3/a_b`
```

`uv init --package` replaces underscore with dash for package name :

```toml
[project]
name = "a-b"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
authors = [
    { name = "Françoise Conil", email = "fcodvpt@gmail.com" }
]
requires-python = ">=3.12"
dependencies = []

[project.scripts]
a-b = "a_b:main"

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
```

I am not a big fan of using underscore, but dash can't be use in import package names.

I am aware of #5255.

Regards
nb : I'm not part of sci-learn team

### Platform

Ubuntu 22.04

### Version

0.6.3

### Python version

Python 3.10.12

**Comment by charliermarsh** [comment]
For clarity, `a-b` and `a_b` are equivalent package names from Python's perspective. Package names have a canonical form, and we use that canonical form everywhere. Notice that https://pypi.org/project/scikit_image/ redirects to https://pypi.org/project/scikit-image/.

We could consider using the verbatim name in `uv init`. I'd prefer not to but am open to more opinions.

**Comment by fconil** [comment]
You're right, I am putting "Name Normalization" documentation link for discussion : <https://packaging.python.org/en/latest/specifications/name-normalization/#name-normalization>

So this is not a bug,  `uv init --package` respects the specifications :/

This is a 2015 specification, and it means we will systematically have different names for "package name" and "import package". I wonder if the security problem could lead to a change but this is more a PyPA question.

I agree to wait for more opinions.

Thanks for your prompt reply

**Comment by amal-khailtash** [comment]
I and a few others have encountered similar issue with some packages that also include C-API.  For example, `CPyCppyy` will be converted to `cpycppyy` which messes with the include directive in one of their C-API files that uses: `include "CPyCppyy/API.h`

Not sure why this normalization is needed?  Shouldn't `uv` be `pip` replaceable?

**Comment by fconil** [comment]
I tried other packaging tools (see below), to see if there was a common behaviour as I did not really get a clear answer from the thread.

Do you think I should ask the Python packaging forum ? 

I am really not convinced to have different names for distribution package and import package.

## Hath 1.14.0

Hatch modifies the name of the distribution package :

```bash
$ hatch new a_b
a-b
├── src
│   └── a_b
│       ├── __about__.py
│       └── __init__.py
├── tests
│   └── __init__.py
├── LICENSE.txt
├── README.md
└── pyproject.toml
```

`pyproject.toml` uses the modified name :

```toml
[project]
name = "a-b"
dynamic = ["version"]
description = ''
```

## Flit 3.11.0

Flit does not generate a skeleton, it can create a `pyproject.toml` files using my answers :

```bash
$ flit init
Module name [a_b]:
Author [F. CONIL]:
…
```

`pyproject.toml` uses the choosen name :

```toml
[project]
name = "a_b"
dynamic = ["version", "description"]
…
```

## Poetry 2.1.1

`poetry new` does not create the base directory

```bash
$ mkdir a_b
$ cd a_b
$ poetry new a_b
Created package a_b in a_b

$ tree a_b
a_b
├── pyproject.toml
├── README.md
├── src
│   └── a_b
│       └── __init__.py
└── tests
    └── __init__.py
```

but it modifies the project name :

```toml
[project]
name = "structify-net"
version = "0.1.0"
description = ""
…
```

## PDM 2.22.4

`pdm init`  does not create the base directory

```bash
$ mkdir a_b
$ cd a_b

$ pdm init
Creating a pyproject.toml for PDM...
Please enter the Python interpreter to use
 0. cpython@3.10 (/usr/bin/python3)
 …
Please select (0): 0
Virtualenv is created successfully at …/PDM-2.22.4/a_b/.venv
Project name (a_b):
Project version (0.1.0):
…
```

`pyproject.toml` uses the choosen name :

```toml
[project]
name = "a_b"
version = "0.1.0"
description = "Network Generation with controlled structure"
```

**Comment by charliermarsh** [comment]
I still think is the right approach, but it's also something that tools can decide for themselves IMO.
